### PR TITLE
Implement slaveExecutorEnvironmentDecorator hook

### PR DESCRIPTION
**Pull request by kamaradclimber**
- implement new hook
- refactored a how we transform protobuf struct->ruby hash

JIRA: JIRA-901
Change-Id: I5f68644838dc0874aacc464e3236b40dc47fa115

### CODE REVIEW

File: RubyHook.cpp [hunk 1]
```diff
@@ -36,32 +60,71 @@ static int unwrapLabels(VALUE key, VALUE value, VALUE arg)
...
   // e.g. in Ruby use: taskInfo["labels"]["foo"] = "bar"
   VALUE labels = rb_hash_new();
   if (taskInfo.has_labels()) {
     foreach (const mesos::Label& l, taskInfo.labels().labels()) {
-      rb_hash_aset(labels, rb_str_new_cstr(l.key().c_str()), rb_str_new_cstr(l.value().c_str()));
+      hash_set(labels, l.key(), l.has_value() ? l.value() : "");
     }
   }
-  rb_hash_aset(hash, rb_str_new_cstr("labels"), labels);
+  hash_set(hash, "labels", labels);
```
Comment (fboismenu) [inline RubyHook.cpp]: ws

File: RubyHook.cpp [hunk 2]
```diff
@@ -9,16 +9,40 @@
...
+namespace {
+  void hash_set(VALUE hash, const std::string& key, const std::string& value) {
+    rb_hash_aset(hash, rb_str_new_cstr(key.c_str()), rb_str_new_cstr(value.c_str()));
+  }
+
+  void hash_set(VALUE hash, const std::string& key, VALUE value) {
+    rb_hash_aset(hash, rb_str_new_cstr(key.c_str()), value);
+  }
+
+  bool hashNotNullOrEmpty(VALUE hash) {
```
Comment (fboismenu) [inline RubyHook.cpp]: name is overkill makes me think (bad!): how about isNonEmptyHash ?

File: RubyHook.cpp [hunk 3]
```diff
@@ -81,6 +144,41 @@ RubyHook::RubyHook(const mesos::Parameters& parameters)
...
+        return Error(ruby.handle_exception());
+      }
+
+      // fill in the Env from Ruby from return value; expect same structure as input
+      if (hashNotNullOrEmpty(result)) {
+        VALUE ruby_cmd = rb_hash_lookup(result, rb_str_new_cstr("command"));
+        if (hashNotNullOrEmpty(ruby_cmd)) {
+          VALUE ruby_env = rb_hash_lookup(ruby_cmd, rb_str_new_cstr("environment"));
+          if (hashNotNullOrEmpty(ruby_env)) {
+            mesos::Environment env;
```
Comment (fboismenu) [inline RubyHook.cpp]: painful ... OPT: 
hash_lookup( VALUE hash, ... )
https://stackoverflow.com/questions/1657883/variable-number-of-arguments-in-c

File: RubyHook.cpp (hunk 2 above)
Comment (kamaradclimber) [reply RubyHook.cpp]: done

File: RubyHook.cpp (hunk 1 above)
Comment (kamaradclimber) [reply RubyHook.cpp]: done

File: RubyHook.cpp (hunk 3 above)
Comment (kamaradclimber) [reply RubyHook.cpp]: I'm not comfortable enough to write that :)

**Pull request merged**